### ISSUE TITLE
Change `GradScaler` to respect an existing `grad_scale` value.

### ISSUE THREAD

**Issue by yousufmo** [pytorch/pytorch#123428]
### 🚀 The feature, motivation and pitch

**Background**
The `grad_scale` value supported by various fused optimizers (e.g.[ Adam](https://github.com/pytorch/pytorch/blob/cfd06bd60cc959622292112ba535a2aadc58f7c7/torch/optim/adam.py#L187) and AdamW) allows a fused kernel to divide the gradients by a denominator before computing the update. The[ `GradScaler` uses this to unscale the gradients](https://github.com/pytorch/pytorch/blob/cfd06bd60cc959622292112ba535a2aadc58f7c7/torch/amp/grad_scaler.py#L435-L437) before computing the optimizer updates in a single kernel. 

**Pitch**
There are conceivable reasons why a user, outside of the `GradScaler` might want to do the same thing. For example, to support efficient custom gradient scaling.

**Existing behaviour**
The issue is the `GradScaler` will override the value of `grad_scale` that a user may have already set. 

**Proposal**
My proposal is simple, if `grad_scale` is already set to `x` and the `GradScaler` has a scale of `y`, we simply set `grad_scale=x*y` before the optimizer step.

In code, this would change 
https://github.com/pytorch/pytorch/blob/cfd06bd60cc959622292112ba535a2aadc58f7c7/torch/amp/grad_scaler.py#L435-L437 
to 
```python
optimizer.grad_scale = (  # type: ignore[attr-defined]
    None if optimizer_state["stage"] == OptState.UNSCALED else scaler * getattr(optimizer, "grad_scale", 1) 
)
``` 
This is safe since post `optimizer.step()` the value gets deleted and so it won't cause the grad_scale to be multiplied again and again.

If this sounds good to you, I've raised a PR [here](https://github.com/pytorch/pytorch/pull/123429)!

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar @mcarilli @ptrblck @leslie-fang-intel @jgong5

**Comment by janeyx99** [comment]
Wait...is this safe? The deletion only happens if the schema already takes in grad_scale, which we're moving away from. So I don't think this is safe!

Taking a step back, it looks like grad_scale and found_inf are not intended to persist/they are only created through the GradScaler. So this would not be the right place to scale grads. What is the custom grad scaling use case in particular that you want? Would clip_grad_norm and using backward hooks (like register_post_accumulate_grad_hook) already satisfy the use case?

**Comment by yousufmo** [comment]
@janeyx99 Many thanks for the response. 

I think it's safe but you're point about the persistence of `grad_scale` is apt.

> **_The deletion only happens if the schema already takes in grad_scale_** 

I believe the deletion happens whenever `not has_grad_scaler_kwarg` holds; which is specifically when the `optimizer.step` does **not** accept the `grad_scaler` in the signature. 

https://github.com/pytorch/pytorch/blob/d59f1da62f6a74d815dede5ac5513b37c49733a3/torch/amp/grad_scaler.py#L410-L412

So I think this is safe.

> **_grad_scale and found_inf are not intended to persist_**

This is a fair point and I recognise this isn't part of the public API really. An alternative is to pass an additional `grad_scale` as an optional argument to `GradScaler.step`. Which would satisfy the use case and I'm happy to put together a PR to do this.

> **_What is the custom grad scaling use case in particular that you want?_**

Great question. The use case is to be able to vertically fuse gradient scaling with the optimizer step. Currently the only possible way to do this is via the `Optimizer.grad_scale` attribute for optimizers that support it (e.g. `Adam(W)` with `fused=True`). This is much faster than having a separate `clip_grad_norm_` followed by `optimizer.step`. 

This avoids gradient scaling being launched as a separate sequence of kernels which is not ideal. 

----

Addendum: there is actually small bug in the original proposal, it should be 

```python
optimizer.grad_scale = (  # type: ignore[attr-defined]
    getattr(optimizer, "grad_scale", None) if optimizer_state["stage"] == OptState.UNSCALED else scaler * getattr(optimizer, "grad_scale", 1) 
)
```
to handle the case where `grad_scaler.unscale_()` has been called before `grad_scaler.step`.

**Comment by janeyx99** [comment]
@yousufmo Hm, for vertical fusion, we recommend using torch.compile. Have you tried doing something (roughly) like:

```
@torch.compile()
def f():
    clip_grad_norm(...)
    optimizer.step()
```

where you would have the default optimizer config (which should run foreach on cuda) and not optimizer(fused=True).

I've never tried the above so it might not fuse as certainly as what you're talking about, but I'm curious how much automatic perf you could already get today without making a change to library code.
